marshal.go

```diff
@@ -266,6 +266,20 @@ func marshalImpl(vo reflect.Value, w io.Writer, pos uint64, parent *Element, opt
 					}
 					pos, err = writeOne(lst[0])
 				}
+			case reflect.Func:
+				val := vn.Call(nil)[0]
+				lst, ok := pealElem(val, e.t == DataTypeBinary, tag.omitEmpty)
+				if !ok {
+					return pos, wrapErrorf(
+						ErrIncompatibleType, "marshalling %s from func", val.Type(),
+					)
+				}
+				if len(lst) != 1 {
+					return pos, wrapErrorf(
+						ErrIncompatibleType, "marshalling %s from func", val.Type(),
+					)
+				}
+				pos, err = writeOne(lst[0])
```

Comment:
I think it's better to call `writeOne` for each element of lst, or allow also the function type `func() (value! hasNext bool)` and receive single value multiple times as similar to the chan version.
(Channel version receives multiple values until closed.)

Reply:
Got it, support slice by f18ac9cfb4abc80c5716cb920efb2c5097d490e5